src/spatial/utils.jl

```diff
@@ -26,191 +22,383 @@ abstract type AbstractSpatialSystem end
+    @inbounds old_rate = hop_rates.rates[species, site]
+    @inbounds hop_rates.rates[species, site] = rate
+    @inbounds hop_rates.sum_rates[site] += rate - old_rate
+    old_rate
+end
+
+function evalhoprate(hop_rates::HopRates1, u, species, site, num_nbs::Int) 
+    @inbounds u[species,site]*hop_rates.hopping_constants[species,site]*num_nbs
+end
+
+############## hopping rates of form L_{s,i,j} ################
+struct HopRates2{F} <: AbstractHopRates
+    hopping_constants::Vector{Matrix{F}} # hopping_constants[i][s,j] is the hopping constant at site i of species s to jth neighbor
+    rates::Vector{Matrix{F}} # rates[i][s,j] is the hopping rate at site i of species s to jth neighbor
+    sum_rates::Vector{F} # sum_rates[i] is the total hopping rate out of site i
 end
 
 """
-initializes SpatialRates with zero rates
+initializes HopRates with zero rates
 """
-function SpatialRates(ma_jumps, num_species, spatial_system::AbstractSpatialSystem)
-    num_sites = num_sites(spatial_system)
-    num_jumps = get_num_majumps(ma_jumps)
-    SpatialRates(num_jumps,num_species,num_sites)
+function HopRates2(hopping_constants::Vector{Matrix{F}}) where F <: Number
+    rates = deepcopy(hopping_constants)
+    map(r -> fill!(r, zero(F)), rates)
+    sum_rates = zeros(F, length(rates))
+    HopRates2{F}(hopping_constants, rates, sum_rates)
 end
 
 """
 make all rates zero
 """
-function reset!(spatial_rates)
-    fill!(spatial_rates.rx_rates, zero(eltype(spatial_rates.rx_rates)))
-    fill!(spatial_rates.hop_rates, zero(eltype(spatial_rates.hop_rates)))
-    fill!(spatial_rates.rx_rates_sum, zero(eltype(spatial_rates.rx_rates_sum)))
-    fill!(spatial_rates.hop_rates_sum, zero(eltype(spatial_rates.hop_rates_sum)))
+function reset!(hop_rates::HopRates2{F}) where F <: Number
+    map(r -> fill!(r, zero(F)), hop_rates.rates)
+    fill!(hop_rates.sum_rates, zero(F))
     nothing
 end
+
 """
-return total reaction rate at site
+sample a reaction at site, return (species, target_site)
 """
-function total_site_rx_rate(spatial_rates, site_id)
-    spatial_rates.rx_rates_sum[site_id]
+function sample_hop_at_site(hop_rates::HopRates2, site, rng, spatial_system) 
+    @inbounds rates_at_site = hop_rates.rates[site]
+    r = rand(rng) * total_site_hop_rate(hop_rates, site)
+    species, n = Tuple(CartesianIndices(rates_at_site)[linear_search(rates_at_site, r)])
+    return species, nth_nbr(spatial_system, site, n)
 end
 
 """
-return total hopping rate out of site
+update rates of single species at site
 """
-function total_site_hop_rate(spatial_rates, site_id)
-    spatial_rates.hop_rates_sum[site_id]
+function update_hop_rate!(hop_rates::HopRates2, species::Int, u, site, spatial_system)
+    rates_at_site = hop_rates.rates[site]
+    @inbounds old_rate = sum(@view rates_at_site[species,:])
+    @inbounds rates_at_site[species,:] = (@view hop_rates.hopping_constants[site][species,:]) * u[species,site]
```

Comment:
Ahh, this is not so great a memory layout since you are slicing across a row and this is a frequently called function. Could it be better to transpose so the layout in this case is site by species?

Reply:
Or we could just drop `old_rate` (also, I completely forgot to actually return it). If some other algorithm needs it, it can be computed before calling `update_hop_rate!`

Reply:
I transposed both `hopping_constants` and `rates` internally (the input is still the same) for better memory access.